Allow unsetting fields when using SSA from CSA

julianvmodesto opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug 
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

When upgrading from CSA to SSA, users expect to be able to unset fields.

Currently, users cannot unset fields when the CSA annotation doesn't conflict with SSA. CSA ends up remaining a co-manager of fields but unsetting fields requires exclusive ownership of a field.

This change looks at the manager CSA annotation, and removes this manager entirely.

#### Which issue(s) this PR fixes:
https://github.com/kubernetes/kubernetes/issues/99003

#### Does this PR introduce a user-facing change?
```release-note
Fix unsetting fields when upgrading from client-side to server-side apply
```

julianvmodesto committed:
Allow unsetting fields when using SSA from CSA

apelisse commented on test/cmd/apply.sh:
```diff
@@ -469,6 +469,38 @@ __EOF__
+  key: value
+  legacy: unused
+__EOF__
+  grep -q "legacy: unused" <<< "$(kubectl get configmap test -o yaml "${kube_flags[@]:?}")" || exit 1
```
Maybe
```suggestion
  grep -q "legacy: unused" <(kubectl get configmap test -o yaml "${kube_flags[@]:?}") || exit 1
```

apelisse reviewed (commented):
The integration test fails with something related to server-side apply, but with a conflict. Not sure if it's related.

apelisse commented:
is this supposed to only add the relevant test? :-)

julianvmodesto commented:
Oh no :) -- I'll try out the idea you mentioned to give the new SSA manager exclusive ownership of the fields: look at the field manager for the last-applied annotation field and remove that manager everywhere.

I'm going to try to fix this test and maybe convert it to an e2e test

apelisse commented:
I think that test looks great and reproduces the error pretty accurately. Though maybe we should have one where the person goes from CSA to SSA *with a change*.

fedebongio commented:
/assign @apelisse 
/triage accepted

apelisse commented:
/wg api-expression

julianvmodesto commented:
PTAL @apelisse.

As discussed, to make sure users upgrading to SSA can unset fields, we remove any manager(s) of the last-applied annotation on the upgrade.

Added a unit test and also added an integration test (not sure if it's necessary w/ the test-cmd already).

I wonder if the removal should be scoped down more, but I think this works.

abagshaw commented:
@julianvmodesto Is there any way we could revive this PR? Having a working transition from CSA -> SSA is a large blocker for moving to SSA for me. Thx!
